aws/resource_aws_elasticsearch_domain.go

```diff
@@ -151,6 +152,29 @@ func resourceAwsElasticSearchDomain() *schema.Resource {
 								elasticsearch.TLSSecurityPolicyPolicyMinTls12201907,
 							}, false),
 						},
+						"custom_endpoint_enabled": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  false,
+						},
+						"custom_endpoint": {
+							Type:     schema.TypeString,
+							Optional: true,
+							StateFunc: func(v interface{}) string {
+								// AWS Provider aws_acm_certification.domain_validation_options.resource_record_name
+								// references (and perhaps others) contain a trailing period, requiring a custom StateFunc
+								// to trim the string to prevent Route53 API error
+								value := strings.TrimSuffix(v.(string), ".")
+								return strings.ToLower(value)
```

Comment:
You're right, I have just lifted that code from `aws/resource_aws_route53_record.go`. I was trying to find a common validation function but it doesn't exist, it might be becose the different AWS API endpoints require different validations to be done. I agree with you about the comment update, I will do that. Thank you!

Reply:
It is generally preferable to leave out these types of `StateFunc` where possible. If the API has issues with trailing periods, we can introduce validation to try and catch that early.

Reply:
Perfect @bflad , I will get rid of that validation.